src/mesh_smoothing.rs

```diff
@@ -0,0 +1,582 @@
+use nalgebra::geometry::Point3;
+
+use crate::convert::cast_u32;
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::vertex_to_vertex_topology;
+
+/// # Mesh relaxation / Laplacian smoothing
+/// Relaxes angles between mesh edges, resulting in a smoother geometry.
+/// The number of vertices, faces and the overall topology remains unchanged.
+/// The more iterations, the smoother result.
+/// Too many iterations may cause slow calculation time.
+///
+/// The algorithm is based on replacing each vertex position
+/// with an average position of its immediate neighbors
+#[allow(dead_code)]
+pub fn laplacian_smoothing(geometry: &Geometry, iterations: u8) -> Geometry {
+    laplacian_smoothing_with_anchors(geometry, iterations, &[])
+}
+
+/// # Mesh relaxation / Laplacian smoothing with fixed vertices
+/// Relaxes angles between mesh edges, while keeping some vertices anchored,
+/// resulting in a smoother geometry.
+/// The number of vertices, faces and the overall topology remains unchanged.
+/// The more iterations, the smoother result.
+/// Too many iterations may cause slow calculation time.
+///
+/// The algorithm is based on replacing each vertex position
+/// with an average position of its immediate neighbors
+#[allow(dead_code)]
+pub fn laplacian_smoothing_with_anchors(
+    geometry: &Geometry,
+    iterations: u8,
+    fixed_vertex_indices: &[u32],
+) -> Geometry {
+    let vertex_to_vertex_topology = vertex_to_vertex_topology(geometry);
+    let mut geometry_vertices: Vec<Point3<f32>> = Vec::from(geometry.vertices());
+    let mut vertices = geometry_vertices.clone();
+
+    for _ in 0..iterations {
+        for (current_vertex_index, neighbors_indices) in vertex_to_vertex_topology.iter() {
+            if fixed_vertex_indices
+                .iter()
+                .all(|i| *i != cast_u32(*current_vertex_index))
+            {
+                let mut average_position: Point3<f32> = Point3::origin();
+                for neighbor_index in neighbors_indices {
+                    average_position += geometry_vertices[*neighbor_index].coords;
+                }
+                average_position /= neighbors_indices.len() as f32;
```

Comment:
Do we hope here the geometry is without orphans? If it has a vertex with no neighbors, this will produce an `inf` value. Not sure what do do, maybe the easiest way is to add an explicit branch and not relax a vertex without neighbors.

Reply:
good point. this actually destroys the orphans. i've changed the logic so that it preserver the orphans just like the fixed vertices